File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/playlist/PlaylistFragment.java
/**
 * @file
 * @brief Playlist fragment component
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.ui.playlist;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.SubMenu;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.selection.ItemDetailsLookup;
import androidx.recyclerview.selection.OnItemActivatedListener;
import androidx.recyclerview.selection.Selection;
import androidx.recyclerview.selection.SelectionPredicates;
import androidx.recyclerview.selection.SelectionTracker;
import androidx.recyclerview.selection.StorageStrategy;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import app.zxtune.Log;
import app.zxtune.R;
import app.zxtune.device.media.MediaSessionModel;
import app.zxtune.playlist.ProviderClient;
import app.zxtune.ui.utils.SelectionUtils;

public class PlaylistFragment extends Fragment {

  private static final String TAG = PlaylistFragment.class.getName();

  private static final String LISTING_STATE_KEY = "listing_state";

  @Nullable
  private ProviderClient ctrl;
  @Nullable
  private RecyclerView listing;
  @Nullable
  private ItemTouchHelper touchHelper;
  @Nullable
  private View stub;
  @Nullable
  private SelectionTracker<Long> selectionTracker;

  public static Fragment createInstance() {
    return new PlaylistFragment();
  }

  @Override
  public void onAttach(Context ctx) {
    super.onAttach(ctx);

    ctrl = ProviderClient.create(ctx);
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setHasOptionsMenu(true);
  }

  @Override
  public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    super.onCreateOptionsMenu(menu, inflater);

    inflater.inflate(R.menu.playlist, menu);
    final MenuItem sortItem = menu.findItem(R.id.action_sort);
    final SubMenu sortMenuRoot = sortItem.getSubMenu();
    for (ProviderClient.SortBy sortBy : ProviderClient.SortBy.values()) {
      for (ProviderClient.SortOrder sortOrder : ProviderClient.SortOrder.values()) {
        try {
          final MenuItem item = sortMenuRoot.add(getMenuTitle(sortBy));
          final ProviderClient.SortBy by = sortBy;
          final ProviderClient.SortOrder order = sortOrder;
          item.setOnMenuItemClickListener(item1 -> {
            ctrl.sort(by, order);
            return true;
          });
          item.setIcon(getMenuIcon(sortOrder));
        } catch (Exception e) {
          Log.w(TAG, e, "onCreateOptionsMenu");
        }
      }
    }
  }

  private static int getMenuTitle(ProviderClient.SortBy by) throws Exception {
    if (ProviderClient.SortBy.title.equals(by)) {
      return R.string.information_title;
    } else if (ProviderClient.SortBy.author.equals(by)) {
      return R.string.information_author;
    } else if (ProviderClient.SortBy.duration.equals(by)) {
      return R.string.statistics_duration;//TODO: extract
    } else {
      throw new Exception("Invalid sort order");
    }
  }

  private static int getMenuIcon(ProviderClient.SortOrder order) throws Exception {
    if (ProviderClient.SortOrder.asc.equals(order)) {
      return android.R.drawable.arrow_up_float;
    } else if (ProviderClient.SortOrder.desc.equals(order)) {
      return android.R.drawable.arrow_down_float;
    } else {
      throw new Exception("Invalid sor order");
    }
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    return processMenu(item.getItemId(), selectionTracker.getSelection()) || super.onOptionsItemSelected(item);
  }

  @Override
  @Nullable
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                           @Nullable Bundle savedInstanceState) {
    return container != null ? inflater.inflate(R.layout.playlist, container, false) : null;
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    listing = view.findViewById(R.id.playlist_content);
    listing.setHasFixedSize(true);

    final ViewAdapter adapter = new ViewAdapter(new AdapterClient());
    listing.setAdapter(adapter);

    selectionTracker = new SelectionTracker.Builder<>("playlist_selection",
        listing,
        new ViewAdapter.KeyProvider(adapter),
        new ViewAdapter.DetailsLookup(listing),
        StorageStrategy.createLongStorage())
        .withSelectionPredicate(SelectionPredicates.<Long>createSelectAnything())
        .withOnItemActivatedListener(new ItemActivatedListener())
        .build();
    adapter.setSelection(selectionTracker.getSelection());

    // another class for test
    final Activity activity = getActivity();
    if (activity instanceof AppCompatActivity) {
      SelectionUtils.install((AppCompatActivity) activity, selectionTracker,
          new SelectionClient());
    }

    if (savedInstanceState != null) {
      restoreState(savedInstanceState);
    }

    final TouchHelperCallback touchHelperCallback = new TouchHelperCallback();
    touchHelper = new ItemTouchHelper(touchHelperCallback);
    touchHelper.attachToRecyclerView(listing);

    stub = view.findViewById(R.id.playlist_stub);

    final Model playlistModel = Model.of(this);
    playlistModel.getItems().observe(getViewLifecycleOwner(), entries -> {
      if (touchHelperCallback.isDragging()) {
        return;
      }
      adapter.submitList(entries);
      if (entries.isEmpty()) {
        listing.setVisibility(View.GONE);
        stub.setVisibility(View.VISIBLE);
      } else {
        listing.setVisibility(View.VISIBLE);
        stub.setVisibility(View.GONE);
      }
    });
    final MediaSessionModel model = MediaSessionModel.of(getActivity());
    model.getState().observe(getViewLifecycleOwner(), state -> {
      final boolean isPlaying = state != null && state.getState() == PlaybackStateCompat.STATE_PLAYING;
      adapter.setIsPlaying(isPlaying);
    });
    model.getMetadata().observe(getViewLifecycleOwner(), metadata -> {
      if (metadata != null) {
        final Uri uri = Uri.parse(metadata.getDescription().getMediaId());
        adapter.setNowPlaying(ProviderClient.findId(uri));
      }
    });
  }

  private void restoreState(Bundle savedInstanceState) {
    selectionTracker.onRestoreInstanceState(savedInstanceState);
    final Parcelable listingState = savedInstanceState.getParcelable(LISTING_STATE_KEY);
    if (listingState != null) {
      listing.getLayoutManager().onRestoreInstanceState(listingState);
    }
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    selectionTracker.onSaveInstanceState(outState);
    outState.putParcelable(LISTING_STATE_KEY, listing.getLayoutManager().onSaveInstanceState());
  }

  private class ItemActivatedListener implements OnItemActivatedListener<Long> {
    @Override
    public boolean onItemActivated(ItemDetailsLookup.ItemDetails<Long> item, MotionEvent e) {
      return onItemClick(item.getSelectionKey());
    }
  }

  private class AdapterClient implements ViewAdapter.Client {
    @Override
    public boolean onDrag(RecyclerView.ViewHolder holder) {
      return onItemDrag(holder);
    }
  }

  private boolean onItemClick(@Nullable Long id) {
    final MediaSessionModel model = MediaSessionModel.of(getActivity());
    final MediaControllerCompat ctrl = model.getMediaController().getValue();
    if (ctrl != null && id != null) {
      final Uri toPlay = ProviderClient.createUri(id);
      ctrl.getTransportControls().playFromUri(toPlay, null);
      return true;
    }
    return false;
  }

  private boolean onItemDrag(RecyclerView.ViewHolder holder) {
    // Disable dragging in selection mode
    if (!selectionTracker.hasSelection()) {
      touchHelper.startDrag(holder);
      return true;
    } else {
      return false;
    }
  }

  // ArchivesService for selection
  private class SelectionClient implements SelectionUtils.Client<Long> {
    @Override
    public String getTitle(int count) {
      return getResources().getQuantityString(R.plurals.tracks,
          count, count);
    }

    @Override
    public List<Long> getAllItems() {
      final RecyclerView.Adapter adapter = listing.getAdapter();
      if (adapter == null) {
        return new ArrayList<>(0);
      }
      final int size = adapter.getItemCount();
      final ArrayList<Long> res = new ArrayList<>(size);
      for (int i = 0; i < size; ++i) {
        res.add(adapter.getItemId(i));
      }
      return res;
    }

    @Override
    public void fillMenu(MenuInflater inflater, Menu menu) {
      inflater.inflate(R.menu.playlist_items, menu);
    }

    @Override
    public boolean processMenu(int itemId, Selection<Long> selection) {
      return PlaylistFragment.this.processMenu(itemId, selection);
    }
  }

  private boolean processMenu(int itemId, Selection<Long> selection) {
    switch (itemId) {
      case R.id.action_clear:
        ctrl.deleteAll();
        break;
      case R.id.action_delete:
        ctrl.delete(convertSelection(selection));
        break;
      case R.id.action_save:
        savePlaylist(convertSelection(selection));
        break;
      case R.id.action_statistics:
        showStatistics(convertSelection(selection));
        break;
      default:
        return false;
    }
    return true;
  }

  @Nullable
  private static long[] convertSelection(@Nullable Selection<Long> selection) {
    if (selection == null || selection.size() == 0) {
      return null;
    }
    final long[] result = new long[selection.size()];
    final Iterator<Long> iter = selection.iterator();
    for (int idx = 0; idx < result.length; ++idx) {
      result[idx] = iter.next();
    }
    return result;
  }

  private void savePlaylist(@Nullable long[] ids) {
    PlaylistSaveFragment.createInstance(ids).show(getFragmentManager(), "save");
  }

  private void showStatistics(@Nullable long[] ids) {
    PlaylistStatisticsFragment.createInstance(ids).show(getFragmentManager(), "statistics");
  }

  private class TouchHelperCallback extends ItemTouchHelper.SimpleCallback {

    boolean isActive = false;
    long draggedItem = -1;
    int dragDelta = 0;

    TouchHelperCallback() {
      super(ItemTouchHelper.UP | ItemTouchHelper.DOWN, 0);
    }

    final boolean isDragging() {
      return isActive;
    }

    @Override
    public boolean isLongPressDragEnabled() {
      return false;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
      return false;
    }

    @Override
    public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
      if (viewHolder != null && actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
        isActive = true;
        viewHolder.itemView.setActivated(true);
      }
      super.onSelectedChanged(viewHolder, actionState);
    }

    @Override
    public void clearView(RecyclerView view, RecyclerView.ViewHolder viewHolder) {
      super.clearView(view, viewHolder);
      isActive = false;
      viewHolder.itemView.setActivated(false);

      if (draggedItem != -1 && dragDelta != 0) {
        ctrl.move(draggedItem, dragDelta);
      }
      draggedItem = -1;
      dragDelta = 0;
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder source,
                          RecyclerView.ViewHolder target) {
      final int srcPos = source.getAdapterPosition();
      final int tgtPos = target.getAdapterPosition();
      if (draggedItem == -1) {
        draggedItem = source.getItemId();
      }
      dragDelta += tgtPos - srcPos;
      final ViewAdapter adapter = ((ViewAdapter) recyclerView.getAdapter());
      if (adapter != null) {
        adapter.onItemMove(srcPos, tgtPos);
      }
      return true;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/playlist/PlaylistFragment.kt
/**
 * @file
 * @brief Playlist fragment component
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.ui.playlist

import android.os.Bundle
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.net.toUri
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.selection.Selection
import androidx.recyclerview.selection.SelectionPredicates
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.selection.StorageStrategy
import androidx.recyclerview.widget.RecyclerView
import app.zxtune.R
import app.zxtune.device.media.MediaModel
import app.zxtune.fs.provider.VfsProviderClient
import app.zxtune.playlist.IdType
import app.zxtune.playlist.Playlist
import app.zxtune.playlist.ProviderClient
import app.zxtune.playlist.Track
import app.zxtune.ui.PersistentStorageSetupFragment
import app.zxtune.ui.utils.SelectionUtils
import app.zxtune.ui.utils.item
import app.zxtune.ui.utils.whenLifecycleStarted
import kotlinx.coroutines.launch

class PlaylistFragment : Fragment() {
    private lateinit var listing: RecyclerView
    private lateinit var search: SearchView
    private lateinit var selectionTracker: SelectionTracker<IdType>

    private val model by activityViewModels<Model>()
    private val mediaModel
        get() = MediaModel.of(requireActivity())
    private val mediaController
        get() = MediaControllerCompat.getMediaController(requireActivity())

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ) = container?.let { inflater.inflate(R.layout.playlist, it, false) }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val panel = view.findViewById<FrameLayout>(R.id.playlist_top_panel)
        setupToolbarMenu(panel)
        listing = setupListing(
            panel, view.findViewById(R.id.playlist_content), view.findViewById(R.id.playlist_stub)
        )
        search = setupSearchView(view.findViewById(R.id.playlist_search))
    }

    private fun setupToolbarMenu(panel: FrameLayout) {
        require(panel.childCount == 1)
        val toolbar = panel.getChildAt(0) as Toolbar
        toolbar.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.playlist, menu)
                // for some reason, onPrepareMenu is not called anymore if menu is shown via
                // showAsAction for some items
                requireNotNull(menu.item(R.id.action_sort).subMenu).run {
                    for (sortBy in Playlist.Sorting.By.entries) {
                        for (sortOrder in Playlist.Sorting.Order.entries) {
                            add(getMenuTitle(sortBy)).run {
                                setOnMenuItemClickListener {
                                    model.sort(Playlist.Sorting(sortBy, sortOrder))
                                    true
                                }
                                setIcon(getMenuIcon(sortOrder))
                            }
                        }
                    }
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem) =
                processMenuItem(menuItem.itemId, selectionTracker.selection)
        })
    }

    private fun setupListing(panel: FrameLayout, listing: RecyclerView, stub: View) =
        listing.apply {
            setHasFixedSize(true)
            val adapter = ViewAdapter(model::reorder).apply {
                adapter = this
            }
            selectionTracker = SelectionTracker.Builder(
                "playlist_selection",
                this,
                ViewAdapter.KeyProvider(adapter),
                ViewAdapter.DetailsLookup(this, adapter),
                StorageStrategy.createLongStorage(), // Way too complex to implement alternative
            ).withSelectionPredicate(SelectionPredicates.createSelectAnything())
                .withOnItemActivatedListener { item, _ ->
                    item.selectionKey?.let { onItemClick(it) }
                    false // for ripple
                }.build().also {
                    adapter.setSelection(it.selection)
                    SelectionUtils.install(panel, it, SelectionClient(adapter))
                }
            viewLifecycleOwner.whenLifecycleStarted {
                launch {
                    model.state.collect { state ->
                        adapter.submitList(state.entries) {
                            if (0 == adapter.itemCount) {
                                visibility = View.GONE
                                stub.visibility = View.VISIBLE
                            } else {
                                visibility = View.VISIBLE
                                stub.visibility = View.GONE
                            }
                        }
                    }
                }
                launch {
                    mediaModel.playbackState.collect { state: PlaybackStateCompat? ->
                        adapter.setIsPlaying(PlaybackStateCompat.STATE_PLAYING == state?.state)
                    }
                }
                launch {
                    mediaModel.metadata.collect { metadata ->
                        metadata?.description?.mediaId?.let {
                            val uri = it.toUri()
                            adapter.setNowPlaying(ProviderClient.findId(uri)?.let(Track::Id))
                        }
                    }
                }
            }
        }

    private fun setupSearchView(view: SearchView) = view.apply {
        isSubmitButtonEnabled = false
        setOnCloseListener {
            post {
                clearFocus()
            }
            false
        }
        setOnQueryTextFocusChangeListener { _, hasFocus ->
            if (!hasFocus && query.isEmpty()) {
                isIconified = true
            }
        }
        setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String) = true
            override fun onQueryTextChange(newText: String): Boolean {
                model.filter = newText
                return true
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        selectionTracker.onSaveInstanceState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        savedInstanceState?.let { state ->
            selectionTracker.onRestoreInstanceState(state)
        }
        model.filter.takeIf { it.isNotEmpty() }?.let { query ->
            search.post {
                search.setQuery(query, false)
            }
        }
    }

    private fun onItemClick(id: IdType) = mediaController?.transportControls?.playFromUri(
        ProviderClient.createUri(id), null
    ) ?: Unit

    // ArchivesService for selection
    private inner class SelectionClient(private val adapter: ViewAdapter) :
        SelectionUtils.Client<IdType> {
        override fun getTitle(count: Int) =
            resources.getQuantityString(R.plurals.tracks, count, count)

        override val allItems
            get() = adapter.currentList.map { it.id.value }

        override fun fillMenu(inflater: MenuInflater, menu: Menu) =
            inflater.inflate(R.menu.playlist_items, menu)

        override fun processMenu(itemId: Int, selection: Selection<IdType>) =
            processMenuItem(itemId, selection)
    }

    private fun processMenuItem(itemId: Int, selection: Selection<IdType>): Boolean {
        when (itemId) {
            R.id.action_clear -> deletionAlert(R.string.delete_all_items_query) {
                model.deleteAll()
            }

            R.id.action_delete -> convertSelection(selection)?.let {
                deletionAlert(R.string.delete_selected_items_query) {
                    model.delete(it)
                }
            }

            R.id.action_save -> savePlaylist(convertSelection(selection))
            R.id.action_statistics -> showStatistics(convertSelection(selection))
            else -> return false
        }
        return true
    }

    // TODO: think about using DialogFragment - complicated lambda passing
    private fun deletionAlert(@StringRes message: Int, action: () -> Unit) =
        AlertDialog.Builder(requireContext()).setTitle(message)
            .setPositiveButton(R.string.delete) { _, _ ->
                action()
            }.show()

    private fun savePlaylist(tracks: Track.IdSet?) = lifecycleScope.launch {
        val persistentStorageSetupAction =
            VfsProviderClient(requireContext()).getNotification("playlists:/".toUri())?.action
        val fragment = persistentStorageSetupAction?.let {
            PersistentStorageSetupFragment.createInstance(it)
        } ?: PlaylistSaveFragment.createInstance(tracks)
        fragment.show(parentFragmentManager, "save")
    }

    private fun showStatistics(tracks: Track.IdSet?) =
        PlaylistStatisticsFragment.createInstance(tracks).show(parentFragmentManager, "statistics")

    companion object {
        @StringRes
        private fun getMenuTitle(by: Playlist.Sorting.By) = when (by) {
            Playlist.Sorting.By.TITLE -> R.string.information_title
            Playlist.Sorting.By.AUTHOR -> R.string.information_author
            Playlist.Sorting.By.DURATION -> R.string.statistics_duration //TODO: extract
        }

        @DrawableRes
        private fun getMenuIcon(order: Playlist.Sorting.Order) = when (order) {
            Playlist.Sorting.Order.ASCENDING -> android.R.drawable.arrow_up_float
            Playlist.Sorting.Order.DESCENDING -> android.R.drawable.arrow_down_float
        }

        @VisibleForTesting
        fun convertSelection(selection: Selection<IdType>) =
            selection.takeUnless { it.isEmpty }?.iterator()?.let { iterator ->
                Track.IdSet(selection.size()) {
                    Track.Id(iterator.next())
                }
            }
    }
}
